### PR TITLE
Output to file, with optional argument. Fixes #91

### LINKED ISSUE #91: Add a keyword for outputting to a file
Add a keyword to the reporter constructors whether to output to the file or the console. The default can be set to the console.

### PR DESCRIPTION
**Pull request by nigef**

### CODE REVIEW

File: python_ta/__init__.py [hunk 1]
```diff
@@ -20,31 +20,36 @@
...
 
-from .reporters import REPORTERS
+from .reporters import REPORTERS, PlainReporter, ColorReporter
 from .patches import patch_all
+from .utils import *
 
 # Local version of website; will be updated later.
 HELP_URL = 'http://www.cs.toronto.edu/~david/pyta/'
-
+DEFAULT_OUTPUT_NAME = '*pyta_output'  # incl star to display first, usually.
```
Comment (david-yz-liu) [inline python_ta/__init__.py]: Don't use `*` in file names.

File: python_ta/__init__.py [hunk 2]
```diff
@@ -20,31 +20,36 @@
...
+OUTPUT_STREAM = sys.stdout
```
Comment (david-yz-liu) [inline python_ta/__init__.py]: Don't alias `sys.stdout`

File: python_ta/__init__.py [hunk 3]
```diff
@@ -67,7 +72,8 @@ def _load_config(linter, config_location):
     linter.read_config_file(config_location)
     linter.config_file = config_location
     linter.load_config_file()
-    print('### Loaded configuration file:', config_location)
+    write_stream('### Loaded configuration file: {}'.format(config_location),
```
Comment (david-yz-liu) [inline python_ta/__init__.py]: Leave this as `print`, as well as every other `print` call in this file.

File: python_ta/__init__.py [hunk 4]
```diff
@@ -96,7 +102,8 @@ def _init_linter(config=None, file_linted=None):
             {'default': 5,
              'type': 'int',
              'metavar': '<number_messages>',
-             'help': 'Display a certain number of messages to the user, without overwhelming them.'}),
+             'help': 'Display a certain number of messages to the user, '
```
Comment (david-yz-liu) [inline python_ta/__init__.py]: Please stop making style changes and revert the ones in this pull request.

File: python_ta/reporters/plain_reporter.py [hunk 5]
```diff
@@ -113,9 +116,20 @@ def sort_messages(self):
             self._sorted_style_messages[msg.msg_id].append(msg)
 
     def show_file_linted(self, filename):
```
Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: This method name should be changed.

File: python_ta/utils.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        pass
+    return open(correct_path, 'a')  # return file object, to append messages to
+
+
+def filename_to_display(filename):
+    """Display the file name, currently consistent with pylint format."""
+    return '{} File: {}'.format('*'*15, filename)
+
+
+def write_stream(string, file_stream):
```
Comment (david-yz-liu) [inline python_ta/utils.py]: This function is unnecessary.

File: python_ta/utils.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    if not os.path.exists(os.path.dirname(correct_path)):
+        raise IOError('path {} does not exist.'.format(output))
+    if os.path.isdir(correct_path):
+        correct_path = os.path.join(correct_path, DEFAULT_OUTPUT_NAME)
+    with open(correct_path, 'w') as _:  # erase file, and close it.
+        pass
+    return open(correct_path, 'a')  # return file object, to append messages to
+
+
+def filename_to_display(filename):
```
Comment (david-yz-liu) [inline python_ta/utils.py]: This belongs in the reporter.

File: python_ta/utils.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    """
+    if not os.path.isdir(rel_path):
+        yield rel_path  # Don't do anything; return the file name.
+    else:
+        for root, _, files in os.walk(rel_path):
+            for filename in (f for f in files if f.endswith('.py')):
+                yield os.path.join(root, filename)  # Format path, from root.
+
+
+def get_file_object(output):
```
Comment (david-yz-liu) [inline python_ta/utils.py]: This belongs in the reporter.

File: python_ta/utils.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import os
+# from sys import stdout
```
Comment (david-yz-liu) [inline python_ta/utils.py]: Remove this line.

File: python_ta/reporters/color_reporter.py [hunk 10]
```diff
@@ -20,11 +20,6 @@ class ColorReporter(PlainReporter):
     def __init__(self, source_lines=None, module_name=''):
         super().__init__(source_lines, module_name)
 
-    def reset_messages(self):
```
Comment (david-yz-liu) [inline python_ta/reporters/color_reporter.py]: Do not make changes that are unrelated to the purpose of this pull request.

File: python_ta/__init__.py [hunk 11]
```diff
@@ -212,44 +214,44 @@ def _get_valid_files_to_check(module_name, local_config):
...
       - list of strings of directories or files -- can have multiple.
       - no argument -- checks the python file containing the function call.
     • `level` is used to specify which checks should be made.
     • `local_config` is a dict of config options or string (config file name).
+    • `output` is an absolute path to capture pyta data output. Default std out.
     """
+    global OUTPUT_STREAM
+    OUTPUT_STREAM = get_file_object(output)  # file object, or std out (default)
+    if OUTPUT_STREAM is not sys.stdout:
+        print('PyTA wrote your output to the file here:', OUTPUT_STREAM.name)
```
Comment (david-yz-liu) [inline python_ta/__init__.py]: This is unnecessary.

**Comment by nigef** [comment]
Thanks @david-yz-liu for the excellent feedback

File: python_ta/reporters/plain_reporter.py [hunk 12]
```diff
@@ -112,10 +117,63 @@ def sort_messages(self):
...
+        """
+        self.current_file = filename
+        print(self.filename_to_display(filename), file=self.out)
+
+        # Augment the reporter with the source code.
+        with open(filename) as f:
+            self._source_lines = [
+                line.rstrip() for line in f.readlines()]
+
+    def get_file_paths(self, rel_path):
```
Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: Move this back to `__init__`. It just doesn't belong in its own file.

File: python_ta/reporters/plain_reporter.py [hunk 13]
```diff
@@ -112,10 +117,63 @@ def sort_messages(self):
...
+        """
+        if not os.path.isdir(rel_path):
+            yield rel_path  # Don't do anything; return the file name.
+        else:
+            for root, _, files in os.walk(rel_path):
+                for filename in (f for f in files if f.endswith('.py')):
+                    yield os.path.join(root, filename)  # Format path, from root.
 
     def print_messages(self, level='all'):
+        """Print the messages for a linted file, outputting code snipped of
```
Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: This change does not belong in this pull request. Also there's a typo on "snippet".

File: python_ta/reporters/plain_reporter.py [hunk 14]
```diff
@@ -112,10 +117,63 @@ def sort_messages(self):
...
+        # Paths may contain system-specific or relative syntax, e.g. `~`, `../`
+        correct_path = os.path.expanduser(output_filepath)
+        if not os.path.exists(os.path.dirname(correct_path)):
+            raise IOError('path {} does not exist.'.format(output_filepath))
+        if os.path.isdir(correct_path):
+            correct_path = os.path.join(correct_path, self._output_file_name)
+        with open(correct_path, 'w') as _:  # erase file, and close it.
+            pass
+        # Use this file object to append messages.
+        self.set_output(open(correct_path, 'a'))
```
Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: This is better, but you should only store the filename, and open the file when absolutely necessary (at the start of `print_messages`). In general, don't store file handles as attributes.

File: python_ta/reporters/plain_reporter.py [hunk 15]
```diff
@@ -112,10 +117,63 @@ def sort_messages(self):
...
+            self.set_output(sys.stdout)
+            return
+        
+        # Paths may contain system-specific or relative syntax, e.g. `~`, `../`
+        correct_path = os.path.expanduser(output_filepath)
+        if not os.path.exists(os.path.dirname(correct_path)):
+            raise IOError('path {} does not exist.'.format(output_filepath))
+        if os.path.isdir(correct_path):
+            correct_path = os.path.join(correct_path, self._output_file_name)
+        with open(correct_path, 'w') as _:  # erase file, and close it.
```
Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: This is unnecessary.

File: python_ta/reporters/plain_reporter.py [hunk 16]
```diff
@@ -69,14 +71,17 @@ class PlainReporter(BaseReporter):
...
+        """
         super().__init__()
         self._error_messages = []
         self._style_messages = []
         self._source_lines = source_lines or []
         self._module_name = module_name
-        self.linter = None
         self._sorted_error_messages = defaultdict(list)
         self._sorted_style_messages = defaultdict(list)
+        self._output_file_name = 'pyta_output'
```
Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: The default here should be `None`. Store the default file name in a global constant.

File: python_ta/__init__.py [hunk 17]
```diff
@@ -212,66 +210,60 @@ def _get_valid_files_to_check(module_name, local_config):
...
+        for locations in _get_valid_files_to_check(current_reporter, module_name, local_config):
+            for file_py in current_reporter.get_file_paths(locations):
                 # Load config file in user location. Construct new linter each
                 # time, so config options don't bleed to unintended files.
-                linter = _init_linter(config=local_config, file_linted=file_py)
+                linter = reset_linter(config=local_config, file_linted=file_py)
                 # The local config may have set a new reporter
-                current_reporter = _init_reporter(linter)
-                current_reporter.show_file_linted(file_py)
+                current_reporter = reset_reporter(linter)
```
Comment (david-yz-liu) [inline python_ta/__init__.py]: I was originally going to say that this function call seems to be missing an argument. But actually, let's say that unlike linters, a new reporter is *not* created for each file, so calling `reset_reporter` should be unnecessary (just register it with the new linter).

File: python_ta/reporters/plain_reporter.py (hunk 15 above)
Comment (nigef) [inline python_ta/reporters/plain_reporter.py]: At the beginning, should we remove the file if it exists, before appending to it?

Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: Basically there should only be one call to `open`, and it should be a "write" call, not "append."

File: python_ta/__init__.py [hunk 18]
```diff
@@ -212,66 +222,60 @@ def _get_valid_files_to_check(module_name, local_config):
...
                     yield filepath
                 else:
-                    current_reporter.show_file_linted(item)
+                    print(reporter.filename_to_display(item))
                     print('Could not find the file called, `{}`\n'.format(item))
             except ImportError:
-                current_reporter.show_file_linted(item)
-                print('Could not find the file called, `{}`\n'.format(item))
+                print(reporter.filename_to_display(item))
+                print('Could not find the file called 2, `{}`\n'.format(item))
```
Comment (david-yz-liu) [inline python_ta/__init__.py]: I don't understand why there's a `2` here.

File: python_ta/reporters/plain_reporter.py [hunk 19]
```diff
@@ -112,8 +118,54 @@ def sort_messages(self):
...
+    def filename_to_display(self, filename):
+        """Display the file name, currently consistent with pylint format."""
+        return '{} File: {}'.format('*'*15, filename)
+
+    def register_file(self, filename):
+        """Display current filename to user. Also does some miscellaneous work 
+        with the file.
+        """
+        self.current_file_linted = filename
+        print(self.filename_to_display(filename), file=self.out)
```
Comment (david-yz-liu) [inline python_ta/reporters/plain_reporter.py]: 1. Don't print the filename here. Print it in `print_messages` as part of the report. That is, `print_messages` should be the only method that actually prints anything.
2. Both `set_output` and the `out` attribute are not necessary. Remove them.
3. `print_messages` should have a local variable that is actually the correct output stream. This way the call to `open` (if any) is as close as possible to the actual call to `print`.
